### PR TITLE
History API

### PR DESCRIPTION
**Pull request by Elyniss**

### CODE REVIEW

**Comment by Elyniss** [comment]
Successful Instrumented CI pipeline: https://github.com/WalletConnect/WalletConnectKotlinV2/actions/runs/5413422505

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/HistoryProtocol.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    private lateinit var relayServerUrl: String
+
+    override fun initialize(relayServerUrl: String) {
+        this.relayServerUrl = relayServerUrl
+    }
+
+    override suspend fun registerTags(tags: List<Tags>, onSuccess: () -> Unit, onError: (Core.Model.Error) -> Unit) {
+        registerTagsUseCase(tags, relayServerUrl).fold(
+            onFailure = { error -> onError(Core.Model.Error(error)) },
+            onSuccess = { onSuccess().also { logger.log("Registered: $tags") } }
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/HistoryProtocol.kt:32]:
```suggestion
            onSuccess = { 
               logger.log("Registered: $tags")
               onSuccess() 
            }
```

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/HistoryProtocol.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        )
+    }
+
+    override suspend fun getMessages(params: MessagesParams, onSuccess: (List<HistoryMessage>) -> Unit, onError: (Core.Model.Error) -> Unit) {
+        getMessagesUseCase(params).fold(
+            onFailure = { error -> onError(Core.Model.Error(error)) },
+            onSuccess = { response ->
+                (response.messages ?: emptyList()).also { messages ->
+                    messages.onEach { request -> historyMessageNotifier.requestsSharedFlow.emit(request.toRelay()) }
+                    onSuccess(messages).also { logger.log("Fetched ${messages.size} messages") }
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/HistoryProtocol.kt:42]: similar to above comment

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/domain/RegisterTagsUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import com.walletconnect.foundation.util.Logger
+
+internal class RegisterTagsUseCase(
+    private val service: HistoryServerService,
+    private val generateJwtStoreClientIdUseCase: GenerateJwtStoreClientIdUseCase,
+    private val historyServerUrl: String,
+    private val logger: Logger,
+) {
+    suspend operator fun invoke(tags: List<Tags>, relayUrl: String): Result<Unit> = runCatching {
+        val authorization = "Bearer ${generateJwtStoreClientIdUseCase(historyServerUrl)}"
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/domain/RegisterTagsUseCase.kt:16]: can we move this to private companion object?

File: androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/network/model/messages/Direction.kt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package com.walletconnect.android.history.network.model.messages
+
+enum class Direction { forward, backward }
```
Comment (TalhaAli00) [inline androidCore/sdk/src/main/kotlin/com/walletconnect/android/history/network/model/messages/Direction.kt:3]: enums are conventionally all caps

**Review by TalhaAli00 — APPROVED** [review]

**Review by jakubuid — APPROVED** [review]
Good work with instrumented tests 👏

**Comment by Elyniss** [comment]
IT: https://github.com/WalletConnect/WalletConnectKotlinV2/actions/runs/5422485641

**Pull request merged**